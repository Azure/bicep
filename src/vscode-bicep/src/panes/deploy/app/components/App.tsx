// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { FC, useState } from "react";
import { VSCodeButton, VSCodeProgressRing } from "@vscode/webview-ui-toolkit/react";
import "./index.css";
import { ParamData } from "../../models";
import { useMessageHandler } from "./hooks/useMessageHandler";
import { WhatIfChangesView } from "./sections/WhatIfChangesView";
import { DeploymentOperationsView } from "./sections/DeploymentOperationsView";
import { DeploymentOutputsView } from "./sections/DeploymentOutputsView";
import { ResultsView } from "./sections/ResultsView";
import { ParametersInputView } from "./sections/ParametersInputView";
import { useAzure } from "./hooks/useAzure";
import { DeploymentScopeInputView } from "./sections/DeploymentScopeInputView";
import { FormSection } from "./sections/FormSection";

export const App: FC = () => {
  const [errorMessage, setErrorMessage] = useState<string>();
  const messages = useMessageHandler({ setErrorMessage });
  const azure = useAzure({
    scope: messages.scope,
    acquireAccessToken: messages.acquireAccessToken,
    templateMetadata: messages.templateMetadata,
    parametersMetadata: messages.paramsMetadata,
    setErrorMessage
  });

  function setParamValue(key: string, data: ParamData) {
    const parameters = Object.assign({}, messages.paramsMetadata.parameters, { [key]: data });
    messages.setParamsMetadata({ ...messages.paramsMetadata, parameters });
  }

  function handleEnableParamEditing() {
    messages.setParamsMetadata({ ...messages.paramsMetadata, sourceFilePath: undefined });
  }

  const azureDisabled = !messages.scope || !messages.templateMetadata || azure.running;

  async function handleDeployClick() {
    messages.publishTelemetry('deployPane/deploy', {});
    await azure.deploy();
  }

  async function handleValidateClick() {
    messages.publishTelemetry('deployPane/validate', {});
    await azure.validate();
  }

  async function handleWhatIfClick() {
    messages.publishTelemetry('deployPane/whatIf', {});
    await azure.whatIf();
  }

  return (
    <main id="webview-body">
      <FormSection title="Experimental Warning">
        <div className="alert-error">
          <span className="codicon codicon-beaker" />
          The Bicep Deployment Pane is an experimental feature.
          <br/>
          Documentation is available <a href="https://github.com/Azure/bicep/blob/main/docs/deploy-ui.md">here</a>. Please raise issues or feature requests <a href={createNewIssueUrl()}>here</a>.
        </div>
      </FormSection>
      <DeploymentScopeInputView
        scope={messages.scope}
        onPickScope={messages.pickScope} />

      <ParametersInputView
        parameters={messages.paramsMetadata}
        template={messages.templateMetadata}
        disabled={azure.running}
        onValueChange={setParamValue}
        onEnableEditing={handleEnableParamEditing}
        onPickParametersFile={messages.pickParamsFile} />

      <FormSection title="Actions">
        {errorMessage && <div className="alert-error">
          <span className="codicon codicon-error" />
          {errorMessage}
        </div>}
        <div className="controls">
          <VSCodeButton onClick={handleDeployClick} disabled={azureDisabled}>Deploy</VSCodeButton>
          <VSCodeButton onClick={handleValidateClick} disabled={azureDisabled}>Validate</VSCodeButton>
          <VSCodeButton onClick={handleWhatIfClick} disabled={azureDisabled}>What-If</VSCodeButton>
        </div>
        {azure.running && <VSCodeProgressRing></VSCodeProgressRing>}
      </FormSection>

      {messages.scope && <>
        <ResultsView result={azure.result} />
        <DeploymentOperationsView scope={messages.scope} operations={azure.operations} />
        <DeploymentOutputsView outputs={azure.outputs} />
        <WhatIfChangesView changes={azure.whatIfChanges} />
      </>}
    </main>
  );
};

function createNewIssueUrl() {
  const title = 'Deployment Pane: <description>'
  const labels = ['story: deploy pane']
  const body = `

`;

  return `https://github.com/Azure/bicep/issues/new?title=${encodeURIComponent(title)}&body=${encodeURIComponent(body)}&labels=${encodeURIComponent(labels.join(','))}`;
}