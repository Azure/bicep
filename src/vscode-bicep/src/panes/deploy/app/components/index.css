#webview-body {
  display: flex;
  flex-direction: column;
}

.form-section {
  display: grid;
  grid-auto-flow: row;
  grid-row-gap: 1rem;
  overflow-x: scroll;
}

.input-row {
  display: grid;
  grid-auto-flow: row;
}

@media screen and (min-width: 1024px) {
  #webview-body {
    padding: 0 8rem;
    max-width: 1024px;
    margin: 0 auto;
  }
}

vscode-button {
  width: fit-content;
  margin-right: 0.5rem;
}