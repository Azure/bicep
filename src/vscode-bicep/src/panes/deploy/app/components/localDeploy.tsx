// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { VSCodeDataGrid, VSCodeDataGridCell, VSCodeDataGridRow } from "@vscode/webview-ui-toolkit/react";
import { FC } from "react";
import { LocalDeploymentOperationContent, LocalDeployResponse } from "../../../../language";
import { FormSection } from "./sections/FormSection";
import { getPreformattedJson } from "./utils";

export const LocalDeployResult: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  const error = result.deployment.error;
  return (
    <FormSection title="Result">
      <p>{result.deployment.provisioningState}</p>
      {error && (
        <VSCodeDataGrid gridTemplateColumns="max-content auto">
          {error && (
            <VSCodeDataGridRow key={0}>
              <VSCodeDataGridCell gridColumn="1">Code</VSCodeDataGridCell>
              <VSCodeDataGridCell gridColumn="2">{error.code}</VSCodeDataGridCell>
            </VSCodeDataGridRow>
          )}
          {error.message && (
            <VSCodeDataGridRow key={1}>
              <VSCodeDataGridCell gridColumn="1">Message</VSCodeDataGridCell>
              <VSCodeDataGridCell gridColumn="2">{error.message}</VSCodeDataGridCell>
            </VSCodeDataGridRow>
          )}
        </VSCodeDataGrid>
      )}
    </FormSection>
  );
};

export const LocalDeployOperations: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.operations) {
    return null;
  }

  return (
    <FormSection title="Operations">
      <VSCodeDataGrid>
        <VSCodeDataGridRow rowType="header">
          <VSCodeDataGridCell gridColumn="1" cellType="columnheader">
            Resource Name
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="2" cellType="columnheader">
            State
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="3" cellType="columnheader">
            Error
          </VSCodeDataGridCell>
        </VSCodeDataGridRow>
        {result.operations.map((operation) => (
          <VSCodeDataGridRow
            key={operation.resourceName}
            style={isFailed(operation) ? { background: "rgba(255, 72, 45, 0.3)" } : {}}
          >
            <VSCodeDataGridCell gridColumn="1">{operation.resourceName}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="2">{operation.provisioningState}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="3">
              {operation.error ? getPreformattedJson(operation.error) : ""}
            </VSCodeDataGridCell>
          </VSCodeDataGridRow>
        ))}
      </VSCodeDataGrid>
    </FormSection>
  );
};

export const LocalDeployOutputs: FC<{ result: LocalDeployResponse }> = ({ result }) => {
  if (!result.deployment.outputs) {
    return null;
  }

  return (
    <FormSection title="Outputs">
      <VSCodeDataGrid>
        <VSCodeDataGridRow rowType="header">
          <VSCodeDataGridCell gridColumn="1" cellType="columnheader">
            Name
          </VSCodeDataGridCell>
          <VSCodeDataGridCell gridColumn="2" cellType="columnheader">
            Value
          </VSCodeDataGridCell>
        </VSCodeDataGridRow>
        {Object.keys(result.deployment.outputs).map((name) => (
          <VSCodeDataGridRow key={name}>
            <VSCodeDataGridCell gridColumn="1">{name}</VSCodeDataGridCell>
            <VSCodeDataGridCell gridColumn="2">
              {getPreformattedJson(result.deployment.outputs[name])}
            </VSCodeDataGridCell>
          </VSCodeDataGridRow>
        ))}
      </VSCodeDataGrid>
    </FormSection>
  );
};

function isFailed(operation: LocalDeploymentOperationContent) {
  return operation.provisioningState.toLowerCase() === "failed";
}
