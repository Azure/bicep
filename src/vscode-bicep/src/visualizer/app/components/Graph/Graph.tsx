// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { useRef, VFC, memo, useCallback, useMemo } from "react";
import cytoscape from "cytoscape";
import styled, { DefaultTheme, withTheme } from "styled-components";

import { createStylesheet } from "./style";
import { createRevealFileRangeMessage } from "../../../messages";
import { vscode } from "../../vscode";
import { useCytoscape } from "../../hooks";
import { CommandBar } from "./CommandBar";

interface GraphProps {
  elements: cytoscape.ElementDefinition[];
  theme: DefaultTheme;
}

const layoutOptions = {
  name: "elk",
  padding: 100,
  fit: true,
  animate: true,
  animationDuration: 800,
  animationEasing:
    "cubic-bezier(0.33, 1, 0.68, 1)" as cytoscape.Css.TransitionTimingFunction,
  elk: {
    algorithm: "layered",
    "layered.layering.strategy": "INTERACTIVE",
    "layered.nodePlacement.bk.fixedAlignment": "BALANCED",
    "layered.cycleBreaking.strategy": "DEPTH_FIRST",
    "elk.direction": "DOWN",
    "elk.aspectRatio": 2.5,
    "spacing.nodeNodeBetweenLayers": 80,
    "spacing.baseValue": 120,
    "spacing.componentComponent": 100,
  },
};

const zoomOptions = {
  minLevel: 0.2,
  maxLevel: 2,
  sensitivity: 0.1,
};

const GraphContainer = styled.div`
  position: absolute;
  left: 0px;
  top: 0px;
  bottom: 0px;
  right: 0px;
  overflow: hidden;
  background-color: ${({ theme }) => theme.canvas.backgroundColor};
  background-image: ${({ theme }) => theme.canvas.backgroundImage};
  background-size: 24px 24px;
  background-position: 12px 12px;
`;

const GraphComponent: VFC<GraphProps> = ({ elements, theme }) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const styleSheet = useMemo(() => createStylesheet(theme), [theme]);
  const [cytoscapeRef, layoutRef] = useCytoscape(elements, styleSheet, {
    containerRef,
    layoutOptions,
    zoomOptions,
    onNodeDoubleTap: useCallback((event: cytoscape.EventObjectNode) => {
      const filePath = event.target.data("filePath");
      const range = event.target.data("range");
      vscode.postMessage(createRevealFileRangeMessage(filePath, range));
    }, []),
  });

  const handleZoomIn = useCallback(() => {
    cytoscapeRef.current?.zoom(cytoscapeRef.current.zoom() + 0.1);
  }, []);

  const handleZoomOut = useCallback(() => {
    cytoscapeRef.current?.zoom(cytoscapeRef.current.zoom() - 0.1);
  }, []);

  const handleLayout = useCallback(() => {
    layoutRef.current?.run();
  }, []);

  const handleFit = useCallback(() => {
    cytoscapeRef.current?.animate(
      {
        fit: {
          eles: cytoscapeRef.current.elements(),
          padding: layoutOptions.padding,
        },
      },
      {
        easing: layoutOptions.animationEasing,
        duration: layoutOptions.animationDuration,
      }
    );
  }, []);

  return (
    <>
      <GraphContainer ref={containerRef} theme={theme} />
      <CommandBar
        onZoomIn={handleZoomIn}
        onZoomOut={handleZoomOut}
        onLayout={handleLayout}
        onFit={handleFit}
      />
    </>
  );
};

export const Graph = memo(
  withTheme(GraphComponent),
  (prevProps, nextProps) =>
    prevProps.theme === nextProps.theme &&
    prevProps.elements.length === nextProps.elements.length &&
    prevProps.elements.every((prevElement, i) => {
      const prevData = prevElement.data;
      const nextData = nextProps.elements[i].data;

      return (
        prevData.id === nextData.id &&
        prevData.hasError === nextData.hasError &&
        prevData.filePath === nextData.filePath &&
        prevData.range?.start?.line === nextData.range?.start?.line &&
        prevData.range?.start?.character === nextData.range?.start?.character &&
        prevData.range?.end?.line === nextData.range?.end?.line &&
        prevData.range?.end?.character === nextData.range?.end?.character &&
        prevData.backgroundDataUri === nextData.backgroundDataUri &&
        prevData.source === nextData.source &&
        prevData.target === nextData.target
      );
    })
);
