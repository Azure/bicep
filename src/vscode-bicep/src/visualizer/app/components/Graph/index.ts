// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
export * from "./Graph";
export * from "./style";
