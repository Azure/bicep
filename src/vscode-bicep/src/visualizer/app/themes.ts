// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { DefaultTheme } from "styled-components";

const fontFamily = getComputedStyle(document.body)
  .getPropertyValue("--vscode-font-family")
  .replace(/"/g, "");

export const darkTheme: DefaultTheme = {
  name: "dark",
  fontFamily,
  common: {
    foregroundColor: "#ffffff",
    foregroundSecondaryColor: "#c1c1c1",
    errorIndicatorColor: "Red",
    errorFreeIndicatorColor: "Green",
  },
  canvas: {
    backgroundColor: "#111111",
    backgroundImage:
      "radial-gradient(circle at 1px 1px, #3f3f3f 1px, transparent 0)",
  },
  graph: {
    childlessNode: {
      backgroundColor: "#333333",
      borderColor: "#c1c1c1",
      borderOpacity: 0.6,
      borderWidth: 1,
    },
    containerNode: {
      backgroundColor: "Black",
      backgroundOpacity: 0.2,
      borderColor: "#c1c1c1",
      borderOpacity: 0.6,
      borderWidth: 1,
    },
    edge: {
      width: 2,
      opacity: 0.6,
      color: "#c1c1c1",
    },
  },
  commandBarButton: {
    backgroundColor: "#333333",
  },
};

export const lightTheme: DefaultTheme = {
  name: "light",
  fontFamily,
  common: {
    foregroundColor: "#323130",
    foregroundSecondaryColor: "#484644",
    errorIndicatorColor: "red",
    errorFreeIndicatorColor: "MediumSpringGreen",
  },
  canvas: {
    backgroundColor: "#ffffff",
    backgroundImage:
      "radial-gradient(circle at 1px 1px, #a0a6af 1px, transparent 0)",
  },
  graph: {
    childlessNode: {
      backgroundColor: "White",
      borderColor: "#484644",
      borderOpacity: 0.6,
      borderWidth: 1,
    },
    containerNode: {
      backgroundColor: "#c0c6cf",
      backgroundOpacity: 0.2,
      borderColor: "#484644",
      borderOpacity: 0.6,
      borderWidth: 1,
    },
    edge: {
      width: 2,
      opacity: 0.6,
      color: "#484644",
    },
  },
  commandBarButton: {
    backgroundColor: "white",
  },
};

export const highContrastTheme: DefaultTheme = {
  name: "high-contrast",
  fontFamily,
  common: {
    foregroundColor: "White",
    foregroundSecondaryColor: "Yellow",
    errorIndicatorColor: "Fuchsia",
    errorFreeIndicatorColor: "Cyan",
  },
  canvas: {
    backgroundColor: "Black",
    backgroundImage: "unset",
  },
  graph: {
    childlessNode: {
      backgroundColor: "#111111",
      borderColor: "Yellow",
      borderOpacity: 1,
      borderWidth: 1,
    },
    containerNode: {
      backgroundColor: "Black",
      backgroundOpacity: 0.2,
      borderColor: "Yellow",
      borderOpacity: 1,
      borderWidth: 1,
    },
    edge: {
      width: 2,
      opacity: 1,
      color: "Yellow",
    },
  },
  commandBarButton: {
    backgroundColor: "black",
  },
};
