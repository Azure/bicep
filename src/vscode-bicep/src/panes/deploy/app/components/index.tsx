// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import ReactDOM from "react-dom";

import { App } from "./App";

ReactDOM.render(
  <App />,
  document.getElementById("root")
);
