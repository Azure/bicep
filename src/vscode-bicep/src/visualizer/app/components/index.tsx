// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import ReactDOM from "react-dom";
import { createGlobalStyle } from "styled-components";

import { App } from "./App";

const GlobalStyle = createGlobalStyle`
  body {
    height: 100vh;
    margin: 0;
    padding: 0;
    display: flex;
    overflow: hidden;
  }

  #root {
    flex: 1 1 auto;
    overflow: hidden;
  }
`;

ReactDOM.render(
  <>
    <GlobalStyle />
    <App />
  </>,
  document.getElementById("root")
);
