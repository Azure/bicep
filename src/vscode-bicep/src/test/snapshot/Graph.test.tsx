// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import "jest-styled-components";

import { Graph } from "../../visualizer/app/components/Graph";
import { darkTheme } from "../../visualizer/app/themes";

describe("component Graph", () => {
  it("should render", () => {
    const graph = renderer.create(
      <ThemeProvider theme={darkTheme}>
        <Graph elements={[]} />
      </ThemeProvider>
    );

    expect(graph.toJSON()).toMatchSnapshot();
  });
});
