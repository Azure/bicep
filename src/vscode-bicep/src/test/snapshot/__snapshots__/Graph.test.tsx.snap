// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Graph should render 1`] = `
Array [
  .c0 {
  position: absolute;
  left: 0px;
  top: 0px;
  bottom: 0px;
  right: 0px;
  overflow: hidden;
  background-color: #111111;
  background-image: radial-gradient(circle at 1px 1px,#3f3f3f 1px,transparent 0);
  background-size: 24px 24px;
  background-position: 12px 12px;
}

<div
    className="c0"
  />,
  .c1 {
  display: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  position: absolute;
  color: white;
  background: black;
  border-radius: 4px;
  padding: 4px 8px;
  margin-top: 2px;
  right: 44px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: none;
}

.c0 {
  position: absolute;
  top: 20px;
  right: 20px;
  z-index: 99999;
  -webkit-filter: drop-shadow(0px 0px 4px rgba(0,0,0,0.25));
  filter: drop-shadow(0px 0px 4px rgba(0,0,0,0.25));
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #333333;
}

.c3:hover {
  -webkit-filter: brightness(125%);
  filter: brightness(125%);
}

.c3:active {
  -webkit-filter: brightness(85%);
  filter: brightness(85%);
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        Zoom in
      </div>
      <div
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="c3"
          onClick={[Function]}
        >
          <svg
            color="#c1c1c1"
            fill="none"
            height={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "#c1c1c1",
              }
            }
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M15.3431 15.2426C17.6863 12.8995 17.6863 9.1005 15.3431 6.75736C13 4.41421 9.20101 4.41421 6.85786 6.75736C4.51472 9.1005 4.51472 12.8995 6.85786 15.2426C9.20101 17.5858 13 17.5858 15.3431 15.2426ZM16.7574 5.34315C19.6425 8.22833 19.8633 12.769 17.4195 15.9075C17.4348 15.921 17.4498 15.9351 17.4645 15.9497L21.7071 20.1924C22.0976 20.5829 22.0976 21.2161 21.7071 21.6066C21.3166 21.9971 20.6834 21.9971 20.2929 21.6066L16.0503 17.364C16.0356 17.3493 16.0215 17.3343 16.008 17.319C12.8695 19.7628 8.32883 19.542 5.44365 16.6569C2.31946 13.5327 2.31946 8.46734 5.44365 5.34315C8.56785 2.21895 13.6332 2.21895 16.7574 5.34315ZM10.1005 7H12.1005V10H15.1005V12H12.1005V15H10.1005V12H7.10052V10H10.1005V7Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        Zoom out
      </div>
      <div
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="c3"
          onClick={[Function]}
        >
          <svg
            color="#c1c1c1"
            fill="none"
            height={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "#c1c1c1",
              }
            }
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M15.3431 15.2426C17.6863 12.8995 17.6863 9.1005 15.3431 6.75736C13 4.41421 9.20101 4.41421 6.85786 6.75736C4.51472 9.1005 4.51472 12.8995 6.85786 15.2426C9.20101 17.5858 13 17.5858 15.3431 15.2426ZM16.7574 5.34315C19.6425 8.22833 19.8633 12.769 17.4195 15.9075C17.4348 15.921 17.4498 15.9351 17.4645 15.9497L21.7071 20.1924C22.0976 20.5829 22.0976 21.2161 21.7071 21.6066C21.3166 21.9971 20.6834 21.9971 20.2929 21.6066L16.0503 17.364C16.0356 17.3493 16.0215 17.3343 16.008 17.319C12.8695 19.7628 8.32883 19.542 5.44365 16.6569C2.31946 13.5327 2.31946 8.46734 5.44365 5.34315C8.56785 2.21895 13.6332 2.21895 16.7574 5.34315ZM7.10052 10V12H15.1005V10L7.10052 10Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        Reset layout
      </div>
      <div
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="c3"
          onClick={[Function]}
        >
          <svg
            color="#c1c1c1"
            fill="currentColor"
            height={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "#c1c1c1",
              }
            }
            viewBox="0 0 16 16"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.38 5h1V4h1V3h-1V2h-1v1h-1v1h1v1zm8 4h-1v1h-1v1h1v1h1v-1h1v-1h-1V9zM14 2V1h-1v1h-1v1h1v1h1V3h1V2h-1zm-2.947 2.442a1.49 1.49 0 0 0-2.12 0l-7.49 7.49a1.49 1.49 0 0 0 0 2.12c.59.59 1.54.59 2.12 0l7.49-7.49c.58-.58.58-1.53 0-2.12zm-8.2 8.9c-.2.2-.51.2-.71 0-.2-.2-.2-.51 0-.71l6.46-6.46.71.71-6.46 6.46zm7.49-7.49l-.32.32-.71-.71.32-.32c.2-.2.51-.2.71 0 .19.2.19.52 0 .71z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        Fit
      </div>
      <div
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="c3"
          onClick={[Function]}
        >
          <svg
            color="#c1c1c1"
            fill="none"
            height={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "#c1c1c1",
              }
            }
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 3H9V5H5V9H3V3Z"
              fill="currentColor"
            />
            <path
              d="M3 21H9V19H5V15H3V21Z"
              fill="currentColor"
            />
            <path
              d="M15 21H21V15H19V19H15V21Z"
              fill="currentColor"
            />
            <path
              d="M21 3H15V5H19V9H21V3Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
]
`;
